xunit-autolabeler-v2/ast_parser/core/cli_yaml.py

```diff
@@ -0,0 +1,302 @@
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from typing import Any, Dict, List, Optional, Tuple
+
+from ast_parser.core import constants
+from ast_parser.lib import file_utils
+
+import yaml
+
+
+def __attr_validate_required_values(
+    yaml_path: str,
+    yaml_entry: Dict[str, Any],
+    tag: str,
+    attr: str
+) -> Optional[str]:
+    """Ensure required-value YAML keys have the correct values
+
+    Some .drift-data.yml keys only allow a predefined set of values.
+
+    Arguments:
+        yaml_path: A path to a .drift-data.yml file
+        yaml_entry: The YAML entry to validate
+        tag: The region tag corresponding to the specified YAML entry
+        attr: The attribute of the YAML entry to validate
+
+    Returns:
+        An error message if an attribute is invalid; None otherwise
+    """
+    if attr and attr in constants.REQUIRED_KEY_VALUES:
+        actual = yaml_entry[attr]
+        expected = constants.REQUIRED_KEY_VALUES[attr]
+        if actual != expected:
+            return (
+                f'Invalid {attr} value in file {yaml_path} '
+                f'for tag {tag}: {actual}, expected {expected} '
+                ' (or omission)')
+    return None
+
+
+def __attr_validate_additions(
+    yaml_path: str,
+    yaml_entry: Dict[str, Any],
+    tag: str,
+    attr: str,
+    grep_tags: List[str]
+):
+    """Ensure additions attributes have correct values
+
+    This function ensures that additions attributes are lists, and that
+    the child tags they refer to are detected in the codebase.
+
+    Arguments:
+        yaml_path: A path to a .drift-data.yml file
+        yaml_entry: The YAML entry to validate
+        tag: The region tag corresponding to the specified YAML entry
+        attr: The attribute of the YAML entry to validate
+        grep_tags: A list of tags existing (not necessarily parsed out of)
+                   the source code
+
+    Returns:
+        An error message if the additions attribute is invalid; None otherwise
```

Comment:
I'd raise `ValueError` instead of returning an error string. Any reason preferring the return value?

Reply:
The overall purpose of this file is to produce an _aggregate list_ of _all_ the errors in a YAML file.

(Think of it like `flake8` - which shows you _everything_ wrong with your code, so you can fix all the errors at once.)

Reply:
I'm sorry I still don't understand.

For example, the error case below:

```Python
    if not isinstance(yaml_entry[attr], list):
        return (f'Additions key for {tag} in '
                f'{yaml_path} is not a list!')
```

Will you directly show this message to the users? What does it mean?

Reply:
Now I realized why I feel uncomfortable seeing this code.
Here are some functions named `validate` and they return None when the target is valid.

Maybe `get_error(s)` instead of `validate`?

Reply:
Done.